src/utils/lockfree.hpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * Canary - A free and open-source MMORPG server emulator
+ * Copyright (©) 2019-2024 OpenTibiaBR <opentibiabr@outlook.com>
+ * Repository: https://github.com/opentibiabr/canary
+ * License: https://github.com/opentibiabr/canary/blob/main/LICENSE
+ * Contributors: https://github.com/opentibiabr/canary/graphs/contributors
+ * Website: https://docs.opentibiabr.com/
+ */
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <vector>
+#include <thread>
+#include <spdlog/spdlog.h>
+
+const int TOTAL_THREADS = static_cast<int>(std::thread::hardware_concurrency());
+const size_t LOCAL_CACHE_LIMIT = std::max(35 / TOTAL_THREADS, 5);
+constexpr size_t STATIC_PREALLOCATION_SIZE = 100;
+
+struct StackNode {
+	void* data;
+	StackNode* next;
+};
+
+template <std::size_t TSize, size_t CAPACITY>
+class LockfreeFreeList {
+public:
+	LockfreeFreeList() :
+		head(nullptr) { }
+
+	static LockfreeFreeList &get() {
+		static LockfreeFreeList instance;
+		return instance;
+	}
+
+	bool pop(void*&result) {
+		StackNode* old_head = head.load(std::memory_order_acquire);
```

Comment:
This could be relaxed, the barrier on the next operation is sufficient as both relate to `head`.